File: server/services/youtube-feed-service/src/main/java/com/richo/reader/youtube_feed_service/YoutubeFeedService.java
package com.richo.reader.youtube_feed_service;

import com.codahale.metrics.MetricRegistry;
import com.codahale.metrics.Timer;
import com.richo.reader.youtube_feed_service.youtube.YoutubeChannelDownloader;
import com.richodemus.reader.common.kafka_adapter.EventStore;
import com.richodemus.reader.dto.FeedId;
import com.richodemus.reader.dto.FeedUrl;
import com.richodemus.reader.events_v2.UserSubscribedToFeed;

import javax.inject.Inject;
import java.util.Optional;

import static com.codahale.metrics.MetricRegistry.name;
import static com.richodemus.reader.events_v2.EventType.USER_SUBSCRIBED_TO_FEED;

public class YoutubeFeedService {
    private final FeedCache cache;
    private final YoutubeChannelDownloader youtubeChannelDownloader;
    private Timer getChannelTimer;

    @Inject
    public YoutubeFeedService(final FeedCache cache,
                              final YoutubeChannelDownloader youtubeChannelDownloader,
                              final MetricRegistry registry,
                              final EventStore eventStore) {
        this.cache = cache;
        this.youtubeChannelDownloader = youtubeChannelDownloader;
        this.getChannelTimer = registry.timer(name(YoutubeFeedService.class, "getChannel"));
        eventStore.consume(event -> {
            if (event.type() == USER_SUBSCRIBED_TO_FEED) {
                final FeedId feedId = ((UserSubscribedToFeed) event).getFeedId();
                registerChannel(feedId);
            }
            return null;
        });
    }

    private void registerChannel(final FeedId feedId) {
        cache.add(feedId);
    }

    public Optional<Feed> getChannel(final FeedId feedId) {
        final Timer.Context context = getChannelTimer.time();
        try {
            return cache.get(feedId);
        } finally {
            context.stop();
        }
    }

    public FeedId getFeedId(final FeedUrl feedUrl) {
        return youtubeChannelDownloader.getFeedId(feedUrl);
    }
}

File: server/services/youtube-feed-service/src/main/kotlin/com/richo/reader/youtube_feed_service/YoutubeFeedService.kt
package com.richo.reader.youtube_feed_service

import com.codahale.metrics.MetricRegistry
import com.codahale.metrics.MetricRegistry.name
import com.richo.reader.youtube_feed_service.youtube.YoutubeChannelDownloader
import com.richodemus.reader.common.kafka_adapter.EventStore
import com.richodemus.reader.dto.FeedId
import com.richodemus.reader.dto.FeedUrl
import com.richodemus.reader.events_v2.EventType.USER_SUBSCRIBED_TO_FEED
import com.richodemus.reader.events_v2.UserSubscribedToFeed
import javax.inject.Inject

class YoutubeFeedService @Inject
internal constructor(private val cache: FeedCache,
            private val youtubeChannelDownloader: YoutubeChannelDownloader,
            registry: MetricRegistry,
            eventStore: EventStore) {
    private val getChannelTimer = registry
            .timer(name(YoutubeFeedService::class.java, "getChannel"))

    init {
        eventStore.consume { event ->
            if (event.type() === USER_SUBSCRIBED_TO_FEED) {
                val feedId = (event as UserSubscribedToFeed).feedId
                registerChannel(feedId)
            }
        }
    }

    private fun registerChannel(feedId: FeedId) {
        cache.add(feedId)
    }

    fun getChannel(feedId: FeedId) = getChannelTimer.time().use {
        cache[feedId]
    }

    fun getFeedId(feedUrl: FeedUrl) = youtubeChannelDownloader.getFeedId(feedUrl)
}
